### PR TITLE
Add scGPT fine tuned 

### PR DESCRIPTION
**Pull request by sainirmayi**
## Describe your changes

A fine-tuned version of scGPT

## Checklist before requesting a review
- [x] I have performed a self-review of my code

- Check the correct box. Does this PR contain:
  - [ ] Breaking changes
  - [x] New functionality
  - [ ] Major changes
  - [ ] Minor changes
  - [ ] Bug fixes

- [ ] Proposed changes are described in the CHANGELOG.md

- [x] CI Tests succeed and look good!

### CODE REVIEW

File: src/methods/scgpt_fine_tuned/config.vsh.yaml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    
+references: 
+  doi: 10.1038/s41592-024-02201-0
+
+links: 
+  documentation: https://scgpt.readthedocs.io/en/latest/
+  repository: https://github.com/bowang-lab/scGPT
+
+info:
+  preferred_normalization: log_cp10k
```
Comment (lazappi) [inline src/methods/scgpt_fine_tuned/config.vsh.yaml:17]: Should we give the normalised data or is it better to give counts and use scGPT's normalisation?

File: src/methods/scgpt_fine_tuned/script.py [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+## VIASH START
+par = {
+  'input_train': 'resources_test/task_label_projection/cxg_immune_cell_atlas/train.h5ad',
+  'input_test': 'resources_test/task_label_projection/cxg_immune_cell_atlas/test.h5ad',
+  'output': 'output.h5ad',
+  'model_name': 'scGPT_human',
+  'model': 'scGPT_human'
+}
+meta = {
+  'name': 'scgpt',
```
Comment (lazappi) [inline src/methods/scgpt_fine_tuned/script.py:29]: I'm not sure it matters but maybe make this match the component name

File: src/methods/scgpt_fine_tuned/config.vsh.yaml (hunk 1 above)
Comment (lazappi) [reply src/methods/scgpt_fine_tuned/config.vsh.yaml:17]: Actually, I think the counts are used anyway

File: src/methods/scgpt_fine_tuned/script.py [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+  log1p=data_is_raw,  # whether to log1p the normalized data
+  result_log1p_key="X_log1p",
+  subset_hvg=False,  # whether to subset the raw data to highly variable genes
+  hvg_flavor="seurat_v3" if data_is_raw else "cell_ranger",
+  binning=hyperparameters["n_bins"],  # whether to bin the raw data and to what number of bins
+  result_binned_key="X_binned",  # the key in input_train.layers to store the binned data
+)
+
+input_test = input_train[input_train.obs["is_test"] == "1"]
+input_train = input_train[input_train.obs["is_test"] == "0"]
```
Comment (lazappi) [inline src/methods/scgpt_fine_tuned/script.py:210]: Is there a reason for doing it this way rather than preprocessing each separately? Is this how the tutorial does it?

File: src/methods/scgpt_fine_tuned/script.py [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+    optimizer_E, schedule_interval, gamma=hyperparameters["schedule_ratio"]
+  )
+  optimizer_D = torch.optim.Adam(discriminator.parameters(), lr=lr_ADV)
+  scheduler_D = torch.optim.lr_scheduler.StepLR(
+    optimizer_D, schedule_interval, gamma=hyperparameters["schedule_ratio"]
+  )
+else:
+  criterion_adv = None
+  optimizer_E = None
+  optimizer_D = None
```
Comment (lazappi) [inline src/methods/scgpt_fine_tuned/script.py:369]: Could some of the optional parts like this be simplified/removed if we only every use one branch?

File: src/workflows/run_benchmark/main.nf [hunk 5]
```diff
@@ -17,7 +17,8 @@ methods = [
   scanvi_scarches,
   seurat_transferdata,
   singler,
-  xgboost
+  xgboost,
+  scgpt_fine_tuned
```
Comment (lazappi) [inline src/workflows/run_benchmark/main.nf:21]: Can we try to keep things in alphabetical order?

**Review by lazappi — COMMENTED** [review]
I had some minor comments/questions but I think this looks pretty good.

File: src/methods/scgpt_fine_tuned/config.vsh.yaml (hunk 1 above)
Comment (sainirmayi) [reply src/methods/scgpt_fine_tuned/config.vsh.yaml:17]: Yeah, it uses counts.

File: src/methods/scgpt_fine_tuned/script.py [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+  raise ValueError(
+    f"scGPT can only be used with human data "
+    f"(dataset_organism is \"{input_train.uns['dataset_organism']}\")"
+  )
+
+if par["model"] is None:
+  print(f"\n>>> Downloading '{par['model_name']}' model...", flush=True)
+  model_drive_ids = {
+    "scGPT_human": "1oWh_-ZRdhtoGQ2Fw24HP41FgLoomVo-y",
+    "scGPT_CP": "1_GROJTzXiAV8HB4imruOTk6PEGuNOcgB",
```
Comment (lazappi) [inline src/methods/scgpt_fine_tuned/script.py:53]: I don't think the CP model is meant to be fine tuned so it can probably be removed (not that it will be used anyway).

File: src/methods/scgpt_fine_tuned/script.py [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+  train_celltype_labels,
+  valid_celltype_labels,
+  train_batch_labels,
+  valid_batch_labels,
+) = train_test_split(
+  all_counts, celltypes_labels, batch_ids, test_size=0.1, shuffle=True
+)
+
+from torchtext.vocab import Vocab
+from torchtext._torchtext import Vocab as VocabPybind
```
Comment (lazappi) [inline src/methods/scgpt_fine_tuned/script.py:235]: Can you please move the imports to the top?

File: src/methods/scgpt_fine_tuned/script.py [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+  cell_emb_style="cls",
+  mvc_decoder_style="inner product",
+  ecs_threshold=hyperparameters["ecs_threshold"],
+  explicit_zero_prob=explicit_zero_prob,
+  use_fast_transformer=hyperparameters["fast_transformer"],
+  fast_transformer_backend="flash", # "linear" or "flash"
+  pre_norm=hyperparameters["pre_norm"],
+)
+
+if model_file is not None:
```
Comment (lazappi) [inline src/methods/scgpt_fine_tuned/script.py:302]: I don't think this `if` is needed, this file should always exist

File: src/methods/scgpt_fine_tuned/script.py [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+scaler = torch.cuda.amp.GradScaler(enabled=hyperparameters["amp"])
+
+
+### Finetune scGPT with task-specific objectives
+
+best_val_loss = float("inf")
+best_avg_bio = 0.0
+best_model = None
+
+import copy
```
Comment (lazappi) [inline src/methods/scgpt_fine_tuned/script.py:346]: Move to the top

File: src/methods/scgpt_fine_tuned/functions.py [hunk 10]
```diff
@@ -0,0 +1,529 @@
...
+                total_dab += loss_dab.item() * len(input_gene_ids) if training_settings["DAB"] else 0.0
+                total_num += len(input_gene_ids)
+
+    if return_raw:
+      return np.concatenate(predictions, axis=0)
+
+    return total_loss / total_num, total_error / total_num
+
+
+def test(
```
Comment (lazappi) [inline src/methods/scgpt_fine_tuned/functions.py:457]: I don't think this function is used anywhere so maybe it could be removed?

**Review by lazappi — CHANGES_REQUESTED** [review]

File: src/methods/scgpt_fine_tuned/script.py (hunk 6 above)
Comment (sainirmayi) [reply src/methods/scgpt_fine_tuned/script.py:53]: Oh yes, I'll remove it. Thanks for noticing!

File: src/methods/scgpt_fine_tuned/script.py (hunk 8 above)
Comment (sainirmayi) [reply src/methods/scgpt_fine_tuned/script.py:302]: You're right, I'll remove it.

File: src/methods/scgpt_fine_tuned/functions.py (hunk 10 above)
Comment (sainirmayi) [reply src/methods/scgpt_fine_tuned/functions.py:457]: This function is used to get the predictions.

**Review by lazappi — APPROVED** [review]
Looks good, apart from the conflicts with `main`.
